src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPathExtensions.cs

```diff
@@ -52,16 +62,86 @@ public static IEdmNavigationSource NavigationSource(this ODataPath path)
         /// <returns>True if the resource if a resource set or collection of primitive or complex types. False otherwise.</returns>
         public static bool IsCollection(this ODataPath path)
         {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
+
             return path.LastSegment.TranslateWith(new IsCollectionTranslator());
         }
 
+        /// <summary>
+        /// Get the string representation of <see cref="ODataPath"/>.
+        /// mainly translate Query Url path.
+        /// </summary>
+        /// <param name="path">Path to perform the computation on.</param>
+        /// <param name="urlKeyDelimiter">Mark whether key is segment</param>
+        /// <returns>The string representation of the Query Url path.</returns>
+        public static string ToResourcePathString(this ODataPath path, ODataUrlKeyDelimiter urlKeyDelimiter)
+        {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
+
+            if (urlKeyDelimiter == null)
+            {
+                throw Error.ArgumentNull(nameof(urlKeyDelimiter));
+            }
+
+            return string.Concat(path.WalkWith(new PathSegmentToResourcePathTranslator(urlKeyDelimiter)).ToArray()).TrimStart('/');

```

Comment:
```suggestion
            return string.Join("/", path.WalkWith(new PathSegmentToResourcePathTranslator(urlKeyDelimiter)));
```

`PathSegmentToResourcePathTranslator` always prepends with `/` when translating a segment, so we end up with the extra `/` that we need to trim. Since it is only used in this method, we should update it to stop prepending this character and then we could greatly reduce the number of string allocations

Reply:
There isn't an easy fix for this since we prepend the `/` when traversing each segment.

Reply:
It looks like this one is already exposed, so I suppose that's fine